site/en/docs/extensions/reference/storage/index.md

```diff
@@ -81,50 +82,115 @@ tubes can be filled, and if they are filled when you put your message in, it get
+to this setting are immediately saved to sync storage by the options page and the background script uses `storage.onChanged` to apply the setting as soon as possible.
 
 ```js
-chrome.storage.onChanged.addListener(function(changes, namespace) {
-  for (var key in changes) {
-    var storageChange = changes[key];
-    console.log('Storage key "%s" in namespace "%s" changed. ' +
-                'Old value was "%s", new value is "%s".',
-                key,
-                namespace,
-                storageChange.oldValue,
-                storageChange.newValue);
+//// options.html ////
+
+<script defer src="options.js"></script>
+<form id="optionsForm">
+  <label for="debug">
+    <input type="checkbox" name="debug" id="debug">
+    Enable debug mode
+  </label>
+</form>
+
+//// options.js ////
+
+// In-page cache of the user's options
+let options = {};
+
+// Initialize the form with the user's option settings
+chrome.storage.sync.get('options', (data) => {
+  Object.assign(options, data?.options);
+  optionsForm.debug.checked = data?.options?.debug;
+});
+
+// Immediately persist options changes
+optionsForm.debug.addEventListener('change', (event) => {
+  options.debug = event.target.checked;
+  chrome.storage.sync.set({options});
+});
+
+//// background.js ////
+
+// Watch for changes to the user's options & apply them
+chrome.storage.onChanged.addListener((changes, area) => {
+  if (area === 'sync' && changes.options) {
+    console.log('enable debug mode?', !!changes.options?.newValue?.debug);
+    setDebugMode(!!changes.options?.newValue?.debug);
   }
 });
 ```
 
-[1]: https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage
-[2]: /docs/extensions/mv2/manifest/incognito
-[3]: /docs/extensions/mv2/manifest/storage
-[4]: /docs/extensions/mv2/tabs
-[5]: #property-sync
-[6]: #property-local
+Since service workers are not always running, Manifest V3 extensions sometimes need to
+asynchronously load data from storage before they execute their event handlers. To do this, the
+below snippet uses an async `action.onClicked` event handler that waits for the `storageCache`
+global to be populated before executing it's logic.
+
+```js
+//// background.js ////
+
+// Where we will expose all the data we retrieve from storage.sync
+let storageCache = {};
+// Asynchronously retrieve data from storage.sync, then cache it
+let initStorageCache = getAllStorageSyncData().then(items => {
+  // Copy the data retrieved from storage into storageCache
+  Object.assign(storageCache, items);
+});
+
+chrome.action.onClicked.addListener(async (tab) => {
+  await initStorageCache;
+  /* normal action handler logic */
+});
+
+// Reads all data out of storage.sync and exposes it via a promise
+//
+// Note: Once the Storage API gains promise support, this function
+// can be greatly simplified.
+function getAllStorageSyncData() {
+  // Immediately return a promise and start asynchronous work
+  return new Promise((resolve, reject) => {
```

Comment:
Why using callbacks for ManifestV3 code if the API returns a Promise inherently?

Reply:
It's called out above. For reasons, this API does not yet support the Promise syntax, see [here](https://chromium.googlesource.com/chromium/chromium/+/master/chrome/common/extensions/api/storage.json). (You'd expect to see `returns_async` keys, rather than callbacks.)

Reply:
The API supports Promise since Chrome 88:
https://chromiumdash.appspot.com/commit/84869dc2d4381fe85fd71d5f41289de26caeb406

Your link shows an old storage.json so maybe there's a bug in chromium.googlesource.com?

This is the current storage.json:
https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/api/storage.json

Reply:
Thanks for calling this out—I did a naïve Google search and the result looked about right so I didn't check that. (I actually don't know what chromium.googlesource.com is meant to be showing then, the last commit on that repo is in 2012.)

Yes, we should update this.

Reply:
While preparing this PR (and retesting before I wrote this) I was not able to a promise back from `chrome.runtime.sync.get()`. I'm planning to file a crbug about this a bit later this evening.

Reply:
Issue filed: https://bugs.chromium.org/p/chromium/issues/detail?id=1197387